### PR TITLE
ci upgrades and half precision for cc >=7.0

### PR DESCRIPTION
**Pull request by cenit**
bonus: openmp support enabled on macOS

edit: half precision for cc >=7.0 already merged in master

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -1,31 +1,49 @@
...
+    #  VCPKG_ROOT: C:\Tools\vcpkg
+    #  VCPKG_DEFAULT_TRIPLET: x64-windows
     - platform: Win64
       USE_CUDA: no
       COMPILER: vs
       configuration: Release
       VCPKG_ROOT: C:\Tools\vcpkg
       VCPKG_DEFAULT_TRIPLET: x64-windows
     - platform: Win64
+      USE_CUDA: no
```
Comment (AlexeyAB) [inline appveyor.yml:41]: Why did you add USE_CUDA: no there?

**Comment by AlexeyAB** [comment]
@cenit Thanks!
So now:
CUDNN_HALF will be availabel for CC 7.0, 7.2, 7.5
CMake >= 3.12 should be used. 
And now it will be tested on MSVS 2019 instead of 2017
Will be tested on ubuntu1804 instead of ubuntu1404
allow_failures for: macOS - clang - vcpkg
openmp support enabled on macOS

**Comment by cenit** [comment]
I am trying to understand what’s wrong with cuda 10.2 on appveyor. I am trying using a newer image to see if it’s better, but looks like it’s not. I will check directly with rdp what’s wrong later, otherwise I will move completely to github actions and that’s it :)
Ubuntu 18.04 was already the one tested, 14.04 was abandoned by us a long time ago but I forgot to install proper cuda version (I kept installing cuda for ubuntu 1404 also on ubuntu 1804, it works but it’s not right)
And yes, now CUDNN half is enabled whatever cc you have, he it 7.0, 7.2 or 7.5

**Comment by AlexeyAB** [comment]
@cenit 

Ok, I will change this line https://github.com/AlexeyAB/darknet/blob/36c73c5b9e3f2e72049fb68566e32632f6c70e85/CMakeLists.txt#L68-L70

to these now
```cmake
    if (NOT "arch=compute_70,code=sm_70" IN_LIST CUDA_ARCH_FLAGS AND NOT "arch=compute_72,code=sm_72" IN_LIST CUDA_ARCH_FLAGS AND NOT "arch=compute_75,code=sm_75" IN_LIST CUDA_ARCH_FLAGS)
      set(ENABLE_CUDNN_HALF "FALSE" CACHE BOOL "Enable CUDNN Half precision" FORCE)
      message(STATUS "Your setup does not supports half precision (it requires CC >= 7.0)")
```

Thus, you can leisurely find the best way for other changes.

**Comment by cenit** [comment]
don't know what's wrong with your appveyor.
mine builds ok all configs for the same commit... https://ci.appveyor.com/project/cenit/darknet
I will implement GitHub actions, they may be more useful now in the long runs if you want to setup a way to automatically release applications from your GitHub repo :)

**Comment by AlexeyAB** [comment]
> I will implement GitHub actions, they may be more useful now in the long runs if you want to setup a way to automatically release applications from your GitHub repo :)

Thanks! How does it work `.github/workflows/ccpp.yml` ?

----

> don't know what's wrong with your appveyor.

The previous commit is failed in the same place in 3 cases - seems to have ended in timeout `1 hour`: https://github.com/AlexeyAB/darknet/pull/5417/commits/a3f6c2b58e8e264175f54ca095dd2431b5732a89

image

----

image

**Comment by cenit** [comment]
I saw that your appveyor is timing out. It's unreasonable to be only there after 1h. Looks like you went into very slow build machines??

**Comment by AlexeyAB** [comment]
> I saw that your appveyor is timing out. It's unreasonable to be only there after 1h. Looks like you went into very slow build machines??

I didn't change anything.

> GitHub actions is the CI/CD system integrated into GitHub now.
> It is much more tightly integrated with the platform, enabling the distribution of also binary compiled files and not just the source code, for each release, for each platform, much more easily than with travis/circleci/appveyor. Also, it enabled code re-use: for example, vcpkg is easier to use in GitHub actions, also cmake, you just have to import modules for each tool you use, instead of having to write the full script by yourself (you can always do that if you want)

It’s good if it is all as easily done as it is written)

**Comment by cenit** [comment]
> I didn't change anything.

I am sure about that. It's just appveyor misbehaving

File: appveyor.yml (hunk 1 above)
Comment (cenit) [reply appveyor.yml:41]: to test configurations with and without Cuda

**Comment by cenit** [comment]
Hi @AlexeyAB 
First of all, sorry for spamming your appveyor CI. It is crawling slow to death and so it is very behind.
The PR is ready for me. All checks already passed on my account.
Technically, now GitHub Actions do all the things that before were split onto multiple CI infrastructures (CircleCI for make, appveyor for windows, travis for linux). I would be for keeping all of them, but if you want after merging this PR you can remove CircleCI, travis and appveyor.
As you can see here https://github.com/cenit/darknet/actions/runs/93107989 , we test many configurations: all make combos you were testing on circleci, linux with and without Cuda, with and without vcpkg, macOS with and without vcpkg, windows with and without cuda, with and without vcpkg, Cygwin and a new entry, MinGW64 (msys2)
Shall we add also Azure Pipelines to your CI infrastructure? 🤣 
(jokes apart, I can do that if you want to expose the absolute portability of your code on all platforms)

**Comment by AlexeyAB** [comment]
@cenit Hi, Thanks!

Yes, I think we should keep both GitHub Actions and CI.

1. Github and Ci(Travis/Appveyor) duplicate each other's work?

2. Does GitHub actions works faster than CIs (Appveyour / Travis)?

3. Can we use shorter names, f.e., `_cm_` instead of `_cmake_`, and `intg_libs` instead of `integrated_libs`? So we will see the full name
image

4. Will Artifacts for all these OSs appear for each Release on GitHub? https://github.com/cenit/darknet/actions/runs/93107989

5. Do we have somewhere a compilation check with cuDNN?

> Shall we add also Azure Pipelines to your CI infrastructure? 

6. What is its (Azure Pipelines) advantage and on which platforms will the compilation be?

**Comment by cenit** [comment]
1. yes
2. GitHub Actions are MUCH faster, at least in my experience. They also run all in parallel, which helps a lot to make them faster. I'd remove appveyor, leaving the others sure. Appveyor is SOOO slow in your case! Please delete the queue, it is still running days after!
3. Ok sure
4. I am not sure I understand the question. You will have the same results as I have in my account, because that branch is the base for this PR. You can then add GitHub integrations to auto-add artifacts to releases (if I am answering your question)
5. On linux, all CUDA setups have CUDNN, with or without vcpkg (nvidia has a repo which is working fine whenever vcpkg is not included in the CI test). On Windows, only those enabled by vcpkg have cudnn (cudnn is easily installed with vcpkg thanks to a ./vcpkg install cudnn, without vcpkg is very difficult to install it because cudnn is behind a login page). The very first GitHub action, in particular, is a 1:1 clone of your CircleCI setup, so also there you have cudnn whenever requested
6. Azure Pipelines is very similar to GitHub Actions, but for now is more powerful when dealing with Pull Requests (you can trigger a rebuild of the pr with just a comment, /azp run). Since your repository in on fire in these days (did you see the explosion of stars?) I assume you might receive a lot of interesting PR. Azure Pipelines might help you triggering rebuilds for example whenever you modify the master, just to check that the PR is always fine

**Comment by AlexeyAB** [comment]
@cenit Thanks! I merged this PR.
